ex/loqui/lib/loqui/client.ex

```diff
@@ -0,0 +1,420 @@
+defmodule Loqui.Client do
+  defmodule State do
+    @max_sequence round(:math.pow(2, 32) - 1)
+
+    alias Loqui.Protocol.{
+      Codec,
+      Codecs,
+      Compressor,
+      Compressors
+    }
+
+    @type timestamp :: pos_integer
+    @type last_ping :: {Loqui.Client.sequence, timestamp}
+    @type t :: %__MODULE__{
+      port: pos_integer,
+      host: char_list,
+      tcp_opts: Loqui.Client.tcp_opts,
+      recv_timeout: pos_integer,
+      loqui_path: String.t,
+      buffer: binary,
+      packet_buffer: iodata,
+      sequence: :go_away | Loqui.Client.sequence,
+      ping_interval: pos_integer,
+      registered_compressors: %{String.t => Compressor.t},
+      registered_codecs: %{String.t => Codec.t},
+      compressor: Compressor.t,
+      codec: Codec.t,
+      waiters: %{Loqui.Client.sequence => pid},
+      last_ping: last_ping
+    }
+
+    defstruct host: nil,
+      port: nil,
+      sock: nil,
+      tcp_opts: nil,
+      connect_timeout: nil,
+      recv_timeout: nil,
+      loqui_path: nil,
+      buffer: <<>>,
+      packet_buffer: [],
+      sequence: 1,
+      ping_interval: nil,
+      registered_compressors: %{},
+      registered_codecs: %{},
+      compressor: Compressors.NoOp,
+      codec: Codecs.Erlpack,
+      waiters: %{},
+      last_ping: nil
+
+    def add_packet(%State{packet_buffer: buffer}=state, packet) do
+      %State{state | packet_buffer: [packet | buffer]}
+    end
+
+    def push_waiter(%State{waiters: waiters}=state, sequence, waiter) do
```

Comment:
I was going for the push/pop thing. Changed to add/remove. 

I disagree that we're recapitulating maps here, we're building a data structure using composition.